### DISCUSSION TITLE
Switching inverters

### DISCUSSION THREAD

**Discussion by harryboer** [ksheumaker/homeassistant-apsystems_ecur#225] [Ideas]
Hi, 
I'm loving the integration into HA!
I'm sure you're all aware of the hidden menu for ECU-R (sunspec) and ECU-C models: http://xxx.xxxx.xxx.xxx/index.php/hidden In this menu there is a function called 'switch_state'. here you can turn inverters on and off. Until recently I thought this was only possible with ECU-C models via adjusting the power output of the inverters. But I tried it with an ECU-R and it works! Now question is that it would be great to add this feature into this HA integration. In the Netherlands we have on certain very sunny days negative energy prices (when on a dynamic energy tariff), you then want to turn off the inverters for those hours. Now it would be great if we could automate that in HA. I'm no programmer so no idea how difficult or even possible this is but love to hear your thoughts on this! I found this on a German forum from someone who worked on this but not integrated into HA, but this might be a starting point? https://www.photovoltaikforum.com/thread/173721-apsystems-wechselrichter-mit-modbus-auslesen-und-beschreiben-yc600-qs1-ds3-usw/?pageNo=3#post3564312

Harry

**Comment by HAEdwin** [comment]
If you can figure out what the full URL would be to change the switch_state for an inverter, I could probably add it. Lowering the grid frequency is the correct way in which the grid operators operate. If the grid threatens to become overloaded, they reduce the grid frequency, causing the inverters to switch off below the tresshold of the under frequency. This is a regulation that applies to all inverters.

**Reply by harryboer** [reply]
The url is: http://xxx.xxx.xxx.xxx/index.php/configuration/switch_state

the login for hidden menu items is ‘admin’ for both user and password. 

In here you can switch individual inverters or switch them all. The page looks like this:
![IMG_0728](https://github.com/ksheumaker/homeassistant-apsystems_ecur/assets/67857523/9ecb2878-91e1-4635-ac80-f5c2c24dc6a6)

**Reply by HAEdwin** [reply]
Sorry, I don't own an ECU-R (SunSpec), so you'll have to help me figure out what command is given when you press the [Turn On/Off inverter] button. I hope you or someone else can figure this out.

**Reply by harryboer** [reply]
Would this help? I copied the source code from the webpage:
when you scroll down there is an item on switching all on and all off.


[ECU source code.docx](https://github.com/ksheumaker/homeassistant-apsystems_ecur/files/13990775/ECU.source.code.docx)

**Reply by harryboer** [reply]
On accessing these menus a username and password is required ('admin' for both) and it times out after a while and then this login window appears again. I also attached the source code of that page if you need this as well.

![Schermafbeelding 2024-01-19 om 19 11 53](https://github.com/ksheumaker/homeassistant-apsystems_ecur/assets/67857523/c2b55d36-087d-4cc1-b36a-09891dcd65ab)
[Login source code.docx](https://github.com/ksheumaker/homeassistant-apsystems_ecur/files/13992785/Login.source.code.docx)

**Reply by HAEdwin** [reply]
I guess it is a toggle with a post to: `http://192.168.2.132/index.php/configuration/set_switch_state?ids=9010000071x` or something like that.

**Reply by harryboer** [reply]
I think that is for only 1 inverter (9010000071x is one of my inverter ID)
Further down is a ‘switch on all’ and ‘switch off all’ command:

```
$("#switch_turn_on_all").click(function(){
 
   ​$("#result").hide();
 
       
 
​    $.ajax({
 
   ​​url : "http://192.168.2.132/index.php/configuration/set_switch_all_on",
 
   ​​type : "post",
 
           dataType : "json",
 
   ​​data: "switch_turn_on_all",
 
   ​​success : function(Results){
 
           ​$("#result").text(Results.message);
 
               if(Results.value == 0){
```

and further a switch off all command:

 ```
$("#switch_turn_off_all").click(function(){
 
   ​$("#result").hide();
 
   ​
 
​    $.ajax({
 
   ​​url : "http://192.168.2.132/index.php/configuration/set_switch_all_off",
 
   ​​type : "post",
 
           dataType : "json",
 
   ​​data: "switch_turn_off_all",
 
   ​​success : function(Results){
 
           ​$("#result").text(Results.message);
 
               if(Results.value == 0){
 

```

**Reply by HAEdwin** [reply]
It's a kind of trial and error,..  I can't test it. Did the first command shut down the specified inverter?

**Reply by harryboer** [reply]
No, Message back was: 
{"value":1,"message":"Please select inverters !"}

I did also try with complete inverter ID, the bit with X replaced with correct digits

but when i used this one all inverters switched off! 

`http://192.168.2.132/index.php/configuration/set_switch_all_off`

With the following command they all switched on again:

`http://192.168.2.132/index.php/configuration/set_switch_all_on`

**Reply by HAEdwin** [reply]
Great, that's not hard to implement. Currently busy with another project but I might squeeze this in this sunday.

**Reply by harryboer** [reply]
That would be great, no rush. There are quite few users I know off that would be very happy. And for my understanding, will it then be an entity that we can change the status of in HA? Like from on to off or send a command?

**Reply by harryboer** [reply]
@HAEdwin Did you mange to find some time to have a look at it?

**Reply by HAEdwin** [reply]
No not yet, still working on another project unfortunately :(

**Reply by HAEdwin** [reply]
@harryboer 
On my git (https://github.com/HAEdwin/homeassistant-apsystems_ecur) there is a first version with the switch "ECU Inverters Online". The idea is that this switch can be turned on/off via an automation.

[off]: the "ECU Inverters Online" switch is turned off and the URL is accessed to turn off the inverters. Check the log and look at the position of the switch.
[on], the "ECU Inverters Online" switch is turned on and the URL is accessed to turn on the inverters. Check the log and look at the position of the switch.

Can you test this? Obviously I get an error on the URL because I own an old ECU-R so it is difficult to test it further. Please provide some logging if you encounter errors.

The automation to shut down the inverters could also contain the use of the ECU_QUERY_DEVICE switch. When inverters are brought offline theoretically queries to the ECU can be stopped.

**Reply by harryboer** [reply]
great, I'll test it in the coming week.

**Reply by harryboer** [reply]
I've tried it today and it works perfectly! All inverter go down when switch is set to 'off' and they come back when I set the switch to 'on'. The inverters don't go off line, but the W go to '0'. The change is instant, when I look at the kWh meter that updates values every second the Power generated goes to (nearly) '0'. After a few minutes when the 5 minute interval updates the ECU values than the values there also drop to '0' in HA. 

I recently installed a few more panels so I now have a mix of QT2 and DS3 inverters. I noticed that the 3 phase wired QT2 (no neutral wire) the voltage goes to 0V, 0W and on the 1 phase DS3 (with null wire) the voltage remains 230 Volts but it does go to 2-3 watts, so almost 0 W. But thats due to the inverters not the integration.

**Reply by HAEdwin** [reply]
Great! Thanks for testing. I can release a beta and see how it turns out for ECU-C owners (might add a check on ECU type as well on the official release).

**Reply by harryboer** [reply]
I think it should work for both ECU-C and the ECU-R 2162 model and not for the ECU-R 2160 model

**Reply by HAEdwin** [reply]
"and on the 1 phase DS3 (with null wire) the voltage remains 230 Volts but it does go to 2-3 watts, so almost 0 W."

I do also have these low outputs on two panels (YC1000), thought it might be the nearby lamppost that generates some power ;)

**Reply by HAEdwin** [reply]
@harryboer Just released v1.4.0 🎉

**Reply by harryboer** [reply]
@HAEdwin Great! Thanks for including it in the integration! 👏🏼

**Comment by patmtp35** [comment]
hi, did you think it was possible to do this for eatch inverters ?
i explain  even if i use a routeur in south of france there is days i export a lot of energy, my idea was to shutt one or 2 inverters during best sun hours .

**Reply by harryboer** [reply]
I don’t know the energy market in France but  why do you want to do this? Is there a limit on how much you can feed back to the grid?

**Comment by patmtp35** [comment]
Normaly WE can t export . But if you don t export more than 1kw . It ok.But in summer Time i could export 5kw a day...So i would liké to limit and shutt some invertersLe 23 mai 2024 09:24, harryboer ***@***.***> a écrit :
I don’t know the energy market in France but  why do you want to do this? Is there a limit on how much you can feed back to the grid?
